### ISSUE TITLE
[error 16] Device or resource busy

### ISSUE THREAD

**Issue by daiqing98** [fossasia/visdom#827]
![3DEA8892-EDC5-4ECF-9FE6-F07F64F9D488](https://user-images.githubusercontent.com/95208297/145459206-619e1e19-e134-45c4-a186-e2d0bf272731.png)

I am not sure if it is because there is something wrong with my code. I understand that I can set a larger ulimit value so that more files can remain open at the same time. But I wonder if it is necessary. My question is that, if there is a way that they are not necessarily opened at the same time? 

I am training my model and drawing a loss function while training.

Many thanks!

**Comment by JackUrb** [comment]
Visdom requires a server process running concurrently with your plotting process to enable live logging. I'm not quite sure this is a file limit issue - your visdom client is unable to connect to the server. Are you able to visit the server (localhost:8097) in your browser? Are there any issues printed in the server script when you're doing this?

**Comment by daiqing98** [comment]
> 

Thanks for your reply. 
<img width="983" alt="Snipaste_2021-12-09_18-20-56" src="https://user-images.githubusercontent.com/95208297/145491057-41f0b5e6-a02a-4bfd-ae36-16b62e035786.png">

Yes. I am able to visit it and it gives me the error during the training.
It also says "too many open files". When I set ulimit = 10000 instead of 1024, it can run for a longer time before it gives the error.

This is how I use it:
`def visualize_scalars(scalars, names, title, iteration, env='main'):
    assert len(scalars) == len(names)
    # Convert scalar tensors to numpy arrays.
    scalars, names = list(scalars), list(names)
    scalars = [s.cpu() if isinstance(s, CUDATensor) else s for s in scalars]
    scalars = [s.numpy() if hasattr(s, 'numpy') else np.array([s]) for s in
               scalars]
    multi = len(scalars) > 1
    num = len(scalars)

    options = dict(
        fillarea=True,
        legend=names,
        width=400,
        height=400,
        xlabel='Iterations',
        ylabel=title,
        title=title,
        marginleft=30,
        marginright=30,
        marginbottom=80,
        margintop=30,
    )

    X = (
        np.column_stack(np.array([iteration] * num)) if multi else
        np.array([iteration] * num)
    )
    Y = np.column_stack(scalars) if multi else scalars[0]

    if title in _WINDOW_CASH:
        _vis(env).line(X=X, Y=Y, win=_WINDOW_CASH[title], opts=options, update='append')
    else:
        _WINDOW_CASH[title] = _vis(env).line(X=X, Y=Y, opts=options)
`
And this is the callback:
`        generator_training_callbacks = [_generator_training_callback(
            loss_log_interval=loss_log_interval,
            image_log_interval=image_log_interval,
            sample_log_interval=sample_log_interval,
            sample_log=sample_log,
            sample_dir=sample_dir,
            sample_size=sample_size,
            current_task=task,
            total_tasks=len(train_datasets),
            total_iterations=generator_iterations,
            batch_size=batch_size,
            replay_mode=replay_mode,
            env=scholar.name,
        )]

def _generator_training_callback(
        loss_log_interval,
        image_log_interval,
        sample_log_interval,
        sample_log,
        sample_dir,
        current_task,
        total_tasks,
        total_iterations,
        batch_size,
        sample_size,
        replay_mode,
        env):

    def cb(generator, progress, batch_index, result):
        iteration = (current_task-1)*total_iterations + batch_index
        progress.set_description((
            '<Training Generator> '
            'task: {task}/{tasks} | '
            'progress: [{trained}/{total}] ({percentage:.0f}%) | '
            'loss => '
            'g: {g_loss:.4} / '
            'c: {c_loss:.4}'
        ).format(
            task=current_task,
            tasks=total_tasks,
            trained=batch_size * batch_index,
            total=batch_size * total_iterations,
            percentage=(100.*batch_index/total_iterations),
            g_loss=result['g_loss'],
            c_loss=-result['c_loss'],
        ))

        # log the losses of the generator.
        
        if iteration % loss_log_interval == 0:
            visual.visualize_scalar(
                result['g_loss'], 'generator g loss', iteration, env=env
            )
            visual.visualize_scalar(
                result['c_loss'], 'generator c loss', iteration, env=env
            )

        # log the generated images of the generator.
        if iteration % image_log_interval == 0:
            visual.visualize_images(
                generator.sample(sample_size).data,
                'generated samples ({replay_mode})'
                .format(replay_mode=replay_mode), env=env,
            )

        # log the sample images of the generator
        if iteration % sample_log_interval == 0 and sample_log:
            utils.test_model(generator, sample_size, os.path.join(
                sample_dir,
                env + '-sample-logs',
                str(iteration)
            ), verbose=False)

    return cb


`

This is how I fire the call backs:
`            for callback in (generator_callbacks or []):
                callback(generator, progress, batch_index, result_g)`

**Comment by daiqing98** [comment]
> Visdom requires a server process running concurrently with your plotting process to enable live logging. I'm not quite sure this is a file limit issue - your visdom client is unable to connect to the server. Are you able to visit the server (localhost:8097) in your browser? Are there any issues printed in the server script when you're doing this?
<img width="983" alt="Snipaste_2021-12-09_18-20-56" src="https://user-images.githubusercontent.com/95208297/145508420-18fc6005-41bf-42dd-b373-039ef33b1668.png">

Thanks for your reply.

Yes. I am able to visit it and it gives me the error during the training.
It also says "too many open files". When I set ulimit = 10000 instead of 1024, it can run for a longer time before it gives the error.

Here are some relevant codes:
1. visual part
2. definition of callbacks
3. fire callbacks

![visual](https://user-images.githubusercontent.com/95208297/145508740-7b70db89-f66e-4b54-b742-d2b19a23c1ae.png)
<img width="555" alt="callbacks" src="https://user-images.githubusercontent.com/95208297/145508742-d7f21a8e-17ae-4f52-8014-7b218c4540fb.png">
<img width="459" alt="fire" src="https://user-images.githubusercontent.com/95208297/145508743-5c6422f4-306c-472a-a501-29dc85074e7e.png">

Many thanks

**Comment by JackUrb** [comment]
What's your `visualize_images` implementation? I suspect it may be leaving open file handles, as nothing here would lead to there being open files.
